### PR TITLE
CFA for dependent parameters typed by generic constraints

### LINKED ISSUE #48345: Control Flow Analysis for Dependent Parameters doesn't work when the parameters are generic
# Bug Report

### 🔎 Search Terms
Control Flow Analysis for Dependent Parameters generic 

### 🕗 Version & Regression Information
v4.7.0-dev.20220302
### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.7.0-dev.20220302#code/C4TwDgpgBAYgrgOwMZQLxQDwBUoQB7AQIAmAzlANoBEAhlQDRQJwC2ARhAE4C6UAPpSpsGUUsE4BLBAHNuAPgAUAOhU1O00gC4oWAJRo5UAG4B7CcQDcAKCtITCMVABmARm3xkaKAoDWU4oxgNCAANiY0xPqohgDeVlAJUBJO3n4kaKjotFT6cYn5UEGh4cRKwCYwEngQxAq6FgkA9I24nJwmnPGJAL5dCcmp-hlZwrl9BUVhEWUmAKpgkJwAwjSkEHUNUM2t7Z35vd0WQA)

### 💻 Code

```ts
type Func = <T extends ["a", number] | ["b", string]>(...args: T) => void;

const f1: Func = (kind, payload) => {
    if (kind === "a") {
        payload.toFixed();  // error
    }
    if (kind === "b") {
        payload.toUpperCase();  // error
    }
};
```

### 🙁 Actual behavior
parameter type not narrowed

### 🙂 Expected behavior
parameter type is narrowed, like in the example from the [typescript 4.6 blog post](https://devblogs.microsoft.com/typescript/announcing-typescript-4-6-beta/#control-flow-analysis-for-dependent-parameters)

```ts
type Func = (...args: ["a", number] | ["b", string]) => void;

const f1: Func = (kind, payload) => {
    if (kind === "a") {
        payload.toFixed();  // 'payload' narrowed to 'number'
    }
    if (kind === "b") {
        payload.toUpperCase();  // 'payload' narrowed to 'string'
    }
};

f1("a", 42);
f1("b", "hello");
```

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #48345.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -25424,7 +25424,7 @@ namespace ts {
                     if (func.parameters.length >= 2 && isContextSensitiveFunctionOrObjectLiteralMethod(func)) {
                         const contextualSignature = getContextualSignature(func);
                         if (contextualSignature && contextualSignature.parameters.length === 1 && signatureHasRestParameter(contextualSignature)) {
-                            const restType = getTypeOfSymbol(contextualSignature.parameters[0]);
+                            const restType = getReducedApparentType(getTypeOfSymbol(contextualSignature.parameters[0]));
```
Comment (Andarist) [inline src/compiler/checker.ts:25427]: A [very easy fix](https://github.com/microsoft/TypeScript/issues/48345#issuecomment-1076982856) indeed 😅

**Comment by ahejlsberg** [comment]
@typescript-bot run dt

**Comment by awerlogus** [comment]
For test suite:

```ts
function a<T extends 'a' | 'b'> (arg: T) {
  if (arg === 'a') {
    // Expected: 'a'
    // Got: T extends "a" | "b"
    arg
  }
}
```

**Review by RyanCavanaugh — APPROVED** [review]

**Pull request merged**